### PR TITLE
Support forwarding OAuth2 authorization header

### LINKED ISSUE #32: Implement OAuth 2
Enable forwarding OAuth tokens and handle them in the backend to pass them into the driver.

### PR DESCRIPTION
**Pull request by kalil-pelissier**
Linked to trinodb/trino-go-client#130:

* Forward authorization header through query args to `trino-go-client` SQL driver to handle it.
* Update `trino-go-client` configuration to handle authorization header:
```
	config := trino.Config{
		ServerURI:        settings.URL.String(),
		Source:           "grafana",
		CustomClientName: "grafana",
          ++    ForwardAuthorizationHeader: "true",
	}
```
close #32

### CODE REVIEW

File: pkg/trino/constants/keys.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package constants
+
+const (
+	AccessTokenKey  = "accessToken"
```
Comment (nineinchnick) [inline pkg/trino/constants/keys.go:4]: No need for this to be in a separate package.

File: pkg/trino/datasource-context.go [hunk 2]
```diff
@@ -45,3 +52,21 @@ func NewDatasource(c sqlds.Driver) *SQLDatasourceWithTrinoUserContext {
...
+	if strings.HasPrefix(header, bearerPrefix) {
+		return strings.TrimPrefix(header, bearerPrefix), true
+	}
+	return "", false
+}
+
+func injectAccessToken(ctx context.Context, req *backend.QueryDataRequest) context.Context {
+	header := req.GetHTTPHeader(backend.OAuthIdentityTokenHeaderName)
+
+	token, ok := extractBearerToken(header)
```
Comment (nineinchnick) [inline pkg/trino/datasource-context.go:66]: Inline this function here.

**Review by nineinchnick — APPROVED** [review]
How did you test this?

**Comment by nineinchnick** [comment]
You can squash all commits and use the PR title as the commit message (describes the why, not what).

File: pkg/trino/datasource.go [hunk 3]
```diff
@@ -79,11 +80,20 @@ func (s *TrinoDatasource) Converters() (sc []sqlutil.Converter) {
 }
 
 func (s *TrinoDatasource) SetQueryArgs(ctx context.Context, headers http.Header) []interface{} {
-	user := ctx.Value("X-Trino-User")
-	if user == nil {
-		return []interface{}{}
+	var args []interface{}
+
+	user := ctx.Value(constants.TrinoUserHeader)
+	accessToken := ctx.Value(constants.AccessTokenKey)
```
Comment (nineinchnick) [inline pkg/trino/datasource.go:86]: This is in the same package now.

File: pkg/trino/datasource-context.go [hunk 4]
```diff
@@ -3,12 +3,20 @@ package trino
...
+	"strings"
+
 	"github.com/grafana/grafana-plugin-sdk-go/backend"
 	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
 	"github.com/grafana/sqlds/v2"
 	"github.com/trinodb/grafana-trino/pkg/trino/models"
 )
 
+const (
+	AccessTokenKey  = "accessToken"
```
Comment (nineinchnick) [inline pkg/trino/datasource-context.go:15]: No need to export these, right?

File: pkg/trino/datasource.go (hunk 3 above)
Comment (kalil-pelissier) [reply pkg/trino/datasource.go]: Yes, something went wrong with the squash.
fixed.

File: pkg/trino/datasource-context.go (hunk 2 above)
Comment (kalil-pelissier) [reply pkg/trino/datasource-context.go]: fixed

**Comment by kalil-pelissier** [comment]
For testing, I ran a local Grafan/Keycloak/Trino stack with docker-compose to check that the authorization header is forwarded between Grafana and Trino.

**Review by nineinchnick — APPROVED** [review]

**Comment by kalil-pelissier** [comment]
Hi @nineinchnick, thanks for approving and merging this PR!
Do you have an estimate on when these changes might be released?

**Comment by nineinchnick** [comment]
I just ran the release and submitted v1.0.8 for veification to Grafana.

**Comment by kalil-pelissier** [comment]
Hi @nineinchnick,
I have noticed that the last release of the plugin hasn't been published on the [official grafana](https://grafana.com/grafana/plugins/trino-datasource/?tab=changelog) repo. How long does it take for it to be published?

**Comment by nineinchnick** [comment]
I've got a review comments from Grafana after I submitted the update, and I haven't had time to address them yet. I wish this was done publicly, like a PR review, but that's the current process. Here's the full message I got:

> We recommend the following changes:
>
> * Instead of wrapping the instance factory function, instead [update your existing function](https://github.com/trinodb/grafana-trino/compare/v1.0.7...v1.0.8#diff-2c471f82eb1caac4a8a6ca3a41d2bc2017b2d2fc21c69115b11b94d588815b96R46) to receive context.Context and [propagate this through](https://github.com/trinodb/grafana-trino/compare/v1.0.7...v1.0.8#diff-8df4c9da431438a569d44c626de34b45218cb9a20f6cd9edf1c7742811ae8866R20) the code paths instead of using context.Background(). This will ensure you code is more forward compatible with any changes to the SDK that rely on context propagation.
> * Please update the README to mention added OAuth forwarding functionality alongside existing auth methods as this will be helpful for users to understand how to make the most out of the plugin.
> 
> Some questions:
> 
> * Considering this update will now pass the access token raw as a SQL query param, can you confirm that this information is not being logged anywhere? I did confirm that the call to log the dsn will only log forwardAuthorizationHeader=true, but just wanted to confirm with you as well.
> * I wanted to check if you found it necessary to pass the OAuth token via context as it should already be available via the headers http.Header argument in the [SetQueryArgs function](https://github.com/trinodb/grafana-trino/compare/v1.0.7...v1.0.8#diff-a83fa9e304bcecdf7689211ef3c1cdab2cb6e3d7e88df49fd1d78c0d8ef65039R81)? I mainly just want to confirm you haven't found a bug.

**Comment by kalil-pelissier** [comment]
Ok, thanks for the update,
How can it be addressed is it an internal process or can we open a PR to fix those reviews?

**Comment by nineinchnick** [comment]
A PR would be very welcome!

**Comment by kalil-pelissier** [comment]
Ok, I will open a PR with the Grafana Team review description and take care of it when I have time this week. 
I'm not sure I get all the points and will probably need some clarification (the first review link looks outdated).

**Comment by nineinchnick** [comment]
Thank you very much! There was only one PR merged after I cut 1.0.8. https://github.com/trinodb/grafana-trino/pull/274/ removed logging of the DSN

**Comment by nineinchnick** [comment]
I just submitted 1.0.9 for review and responded to all the questions I got for 1.0.8.

**Comment by kalil-pelissier** [comment]
Hi @nineinchnick,
Is there any update for the publication of v1.0.9?

**Comment by nineinchnick** [comment]
No, I need to test and merge #277 before I can submit another update for review.

**Pull request merged**